src/machine.rs

```diff
@@ -0,0 +1,140 @@
+    csrw!(mcounteren, 0xffffffff);
+    csrw!(mscratch, M_MODE_STACK_BASE + M_MODE_STACK_STRIDE * hartid);
+
+    csrw!(pmpaddr7, 0xffffffff_ffffffff);
+    csrs!(pmpcfg0, 0x1f << 56);
+
+    if SHARED_STATICS.hart_lottery.swap(false,  Ordering::SeqCst) {
+        asm!("LOAD_ADDRESS t0, mtrap_entry
+              csrw 0x305, t0 // mtvec"
+             ::: "t0"  : "volatile");
+
+        // Text segment
+        pmp::install_pmp_napot(0, pmp::LOCK | pmp::READ | pmp::EXEC, 0x80000000, 0x200000);
+        // Shared data segment
+        pmp::install_pmp_napot(1, pmp::LOCK | pmp::READ | pmp::WRITE, 0x80200000, 0x200000);
+
+        // Minimal page table to boot into S mode. See [1] for FU540 errata related to mixing huge
+        // pages and PMP.
+        //
+        // [1] https://github.com/riscv/riscv-isa-manual/issues/347
+        let boot_page_table_pa = SHARED_STATICS.boot_page_table.as_ptr() as u64;
+        *((boot_page_table_pa) as *mut u64) = 0x00000000 | 0xcf;
+        *((boot_page_table_pa+16) as *mut u64) = ((boot_page_table_pa + 4096) >> 2) | 0x01;
+        *((boot_page_table_pa+24) as *mut u64) = 0x30000000 | 0xcf;
+        *((boot_page_table_pa+4088) as *mut u64) = ((boot_page_table_pa + 4096) >> 2) | 0x01;
+        *((boot_page_table_pa+4096) as *mut u64) = 0x20000000 | 0xcb;
+        for i in 1..512 {
+            *((boot_page_table_pa + 4096 + i*8) as *mut u64) = (0x20000000 + (i<<19)) | 0xc7;
+        }
+        csrw!(satp, 8 << 60 | (boot_page_table_pa >> 12));
+
+        machdebug::machine_debug_puts("Hello");
+        // pmp::debug_pmp();
+        // pagedebug::debug_paging();
+
+        // TODO: figure out why we have to do this dance instead of just assigning things directly
+        // i.e. why is it that rust will assign a0/a1? how do we stop that? In the mean time, use
+        // the `gp` and `tp` registers at temporaries (the ABI prohibits Rust from passing arguments
+        // in them).
+        asm!("mv gp, $1
+              mv tp, $0
+              mv a0, gp
+              mv a1, tp
+              mret" :: "r"(device_tree_blob), "r"(hartid) : "a0", "a1", "gp", "tp" : "volatile");
+    } else  {
+        asm!("LOAD_ADDRESS t0, start_hart
+             csrw 0x305, t0 // mtvec"
+             ::: "t0"  : "volatile");
+        csrsi!(mstatus, 0x8); //MIE
+        loop {}
+    }
+}
+
+#[link_section = ".text.init"]
+#[no_mangle]
+pub unsafe fn handle_ipi() {
+    let hartid = csrr!(mhartid);
+    let reason = { SHARED_STATICS.ipi_reason_array.get_unchecked(hartid as usize).lock().take() };
+
+    match reason {
+        Some(Reason::EnterSupervisor{ a0, a1, a2, a3, sp, satp, mepc}) => {
+            csrw!(mepc, mepc);
+            csrw!(satp, satp);
+            asm!("mv a0, $0
+                  mv a1, $1
+                  mv a2, $2
+                  mv a3, $3
+                  mv sp, $4
+                  mret" :: "r"(a0), "r"(a1), "r"(a2), "r"(a3), "r"(sp) : "a0", "a1", "a2", "a3", "sp" : "volatile");
```

Comment:
I'm a little worried about what happens here if rust assigns registers differently than we expect, and these mv's all clobber each other.

Reply:
I finally took the time to report a bug based on that so hopefully it gets fixed upstream: https://github.com/rust-lang/rust/issues/60392.